src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -797,35 +797,45 @@ fn is_pure_fn_except_for_input_ref(expr: &ExprImpl) -> bool {
     }
 }
 
-// We are trying to derive a predicate to apply to the other side of a join if all
-// the `InputRef`s in the predicate are eq condition columns, and can hence be substituted
-// with the corresponding eq condition columns of the other side.
-//
-// Strategy:
-// 1. If the function is pure except for any `InputRef`, then we proceed. Else abort.
-// 2. Then, we collect `InputRef`s in the conjunction.
-// 3. If they are all columns in the given side of join eq condition, then we proceed. Else abort.
-// 4. We then rewrite the `ExprImpl`, by replacing `InputRef` column indices with
-//    the equivalent in the other side.
+/// We are trying to derive a predicate to apply to the other side of a join if all
+/// the `InputRef`s in the predicate are eq condition columns, and can hence be substituted
+/// with the corresponding eq condition columns of the other side.
+///
+/// Strategy:
+/// 1. If the function is pure except for any `InputRef` (which may refer to impure computation),
+///    then we proceed. Else abort.
+/// 2. Then, we collect `InputRef`s in the conjunction.
+/// 3. If they are all columns in the given side of join eq condition, then we proceed. Else abort.
+/// 4. We then rewrite the `ExprImpl`, by replacing `InputRef` column indices with
+///    the equivalent in the other side.
+///
+/// # Arguments
+///
+/// Suppose we derive a predicate from the left side to be pushed to the right side.
+/// * `expr`: An expr from the left side.
+/// * `col_num`: The number of columns in the left side.
+/// * `other_size_schema`: The right side's schema.
 fn derive_predicate_from_eq_condition(
     expr: &ExprImpl,
     eq_condition: &EqJoinPredicate,
-    cols_num: usize,
+    col_num: usize,
+    other_size_schema: &Schema,
     expr_is_left: bool,
 ) -> Option<ExprImpl> {
```

Comment:
IIRC, `EqJoinPredicate`'s `eq_keys` already contains all the datatype (both sides) we need. We just need to change the mapping  `HashMap<usize, usize>`  into `HashMap<usize, (usize, DataType)>`.  We don't need another parameter `other_size_schema` (typo) actually.